<a href="https://colab.research.google.com/github/zzhenxi/study-NLP-with-PyTorch/blob/main/%5BChapter3%5D_%EC%8B%A0%EA%B2%BD%EB%A7%9D%EC%9D%98_%EA%B8%B0%EB%B3%B8_%EA%B5%AC%EC%84%B1_%EC%9A%94%EC%86%8C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [15]:
import torch
import torch.nn as nn

'''
note 
클래스 상속과 super에 대해서

여기서 Perceptron이라는 class는 nn.Module을 상속 받았다. 
부모 : nn.Module (잠깐! torch.nn.module은 pytorch에서 모든 신경망 모델에 대한 베이스 class이며 신경망 모델을 다룰때 꼭! subclassing 해줘야 한다고 한다.)
자식 : Perceptron

만약 자식에게 __init__이 없는 경우라면, 
-> 부모인 nn.Module에서 생성자 함수가 생긴다. 따라서 부모의 생성자 함수의 내용을 출력 가능하다.

하지만 여기에서는 자식의 __init__이 있다. 여기서 부모의 생성자 함수의 내용을 쓰려면? 
-> super().__init__(self)을 추가해주면 된다. 

그런데 여기에서는 super(Perceptron, self).__init__()라고 써주었다. 그 이유는? 
-> python 3. 부터 바뀐 방식 https://seeyoulater9468.tistory.com/155

'''

class Perceptron(nn.Module):
  def __init__(self, input_dim):
    super(Perceptron, self).__init__()
    self.fc1 = nn.Linear(input_dim, 1) # 이 Linear에서 w, b를 관리한다. 
  
  def forward(self, x_in):
    # x_in : 입력 데이터 텐서 (batch, num_features)
    return torch.sigmoid(self.fc1(x_in)).squeeze() # (batch,)

In [4]:
# 평균 제곱 오차 손실
import torch
import torch.nn as nn

mse_loss = nn.MSELoss()
outputs = torch.randn(3,5, requires_grad=True)
targets = torch.randn(3,5)
loss = mse_loss(outputs, targets)
print(loss)

tensor(1.3589, grad_fn=<MseLossBackward0>)


In [5]:
outputs

tensor([[-1.2099, -1.3108,  0.4623, -0.1381, -1.6445],
        [ 0.8303,  0.2095, -1.1764, -0.1884,  1.7156],
        [-0.5855,  0.3423, -0.3706,  0.5491,  0.0103]], requires_grad=True)

In [6]:
targets

tensor([[-1.1403, -1.2032, -1.4843,  0.3189,  0.6651],
        [ 1.0640, -0.2253, -0.3024, -0.5579, -0.9045],
        [-0.7953, -0.1583,  0.7137,  1.7073,  0.4726]])

In [1]:
# 크로스 엔트로피 손실 
import torch 
import torch.nn as nn

ce_loss = nn.CrossEntropyLoss()
outputs = torch.randn(3,5, requires_grad=True)
targets = torch.tensor([1,0,3], dtype=torch.int64)
loss = ce_loss(outputs, targets)
print(loss)


tensor(1.7372, grad_fn=<NllLossBackward0>)


In [2]:
outputs

tensor([[-0.3283, -0.0957,  0.6089,  0.2625, -0.9080],
        [-0.2483,  0.5140, -0.4746, -0.8251, -0.0808],
        [-0.7628, -0.0043,  1.1632,  0.4296,  0.9215]], requires_grad=True)

In [3]:
targets

tensor([1, 0, 3])

미니 배치 1 : [-0.3283, -0.0957,  0.6089,  0.2625, -0.9080]   
미니 배치 2 : [-0.2483,  0.5140, -0.4746, -0.8251, -0.0808]   
미니 배치 3 : [-0.7628, -0.0043,  1.1632,  0.4296,  0.9215]    
미니 배치 1에 대한 타겟값 : 1   
미니 배치 2에 대한 타겟값 : 0   
미니 배치 3에 대한 타겟값 : 3   

미니배치를 뜯어보면, [-0.3283, -0.0957,  0.6089,  0.2625, -0.9080] 은 각각 0, 1, 2, 3, 4에 대한 아웃풋이다. 여기서는 2번 값이 가장 크다. 그러니 소프트맥스를 하면 1에 대한 가장 큰 비율을 차지 할 것이다. 

In [9]:
# 이진 크로스 엔트로피 손실

bce_loss = nn.BCELoss()
sigmoid = nn.Sigmoid()
probabilities = sigmoid(torch.randn(4,1, requires_grad=True)) #sigmoid가 0-1사이의 범위로 압축해줌
targets = torch.tensor([1,0,1,0], dtype=torch.float32).view(4,1)
loss = bce_loss(probabilities, targets)

In [10]:
probabilities

tensor([[0.5504],
        [0.7669],
        [0.6396],
        [0.4145]], grad_fn=<SigmoidBackward0>)

In [11]:
targets

tensor([[1.],
        [0.],
        [1.],
        [0.]])

In [12]:
loss

tensor(0.7590, grad_fn=<BinaryCrossEntropyBackward0>)

In [16]:
import torch.optim as optim 
import torch.nn as nn

input_dim = 2
lr = 0.001

perceptron = Perceptron(input_dim=input_dim)
bce_loss = nn.BCELoss()
optimizer = optim.Adam(params=perceptron.parameters(), lr=lr)
